File: src/main/java/lmller/github/io/gofkotlin/iterator/Iterator.java
package lmller.github.io.gofkotlin.iterator;

import java.util.Arrays;
import java.util.List;

public class Iterator {
    public static class Sentence {
        private final List<String> words;

        public Sentence(List<String> words) {
            this.words = words;
        }

        public List<String> getWords() {
            return words;
        }
    }

    public static class IterableSentence extends Sentence implements Iterable<String> {

        public IterableSentence(List<String> words) {
            super(words);
        }

        @Override
        public java.util.Iterator<String> iterator() {
            return getWords().iterator();
        }
    }

    public static void main(String[] args) {
        String[] sentence = "This is not the greatest example in the world".split(" ");
        List<String> words = Arrays.asList(sentence);

        for (String s1 : new Sentence(words).getWords()) {
            //not so cool
        }

        for (String s1 : new IterableSentence(words)) {
            //better
        }

    }
}

File: src/main/java/lmller/github/io/gofkotlin/iterator/Iterator.kt
package lmller.github.io.gofkotlin.iterator

import java.util.*
import kotlin.collections.Iterator

class Sentence(val words: List<String>)
class OtherSentence(private val words: List<String>): Iterable<String> by words

operator fun Sentence.iterator(): Iterator<String> = words.iterator()

fun main(args: Array<String>) {
    val words = "This is not the greatest example in the world"
            .split(" ")
            .toList()

    val words2 = "This is just a tribute"
            .split(" ")
            .toList()

    for(word in Sentence(words)){
        print("$word ")
    }
    println()

    val otherSentence = OtherSentence(words2)
    otherSentence.forEach { word ->
        print("$word ")
    }

    println(otherSentence.contains("tribute"))

}